### PR TITLE
Pandroid: GlobalConfig and create node data storage

### PR DESCRIPTION
**Pull request by GabrielBRDeveloper**
Classes for manager Global preferences and node data storage:

## About this

I need some thing for dynamic store and manage data/preferences in **pandroid**

GlobalConfig: Manage all global preferences from the app.

NodeObject/NodeArray: some thing for store and parse data with event support.

### CODE REVIEW

**Comment by GabrielBRDeveloper** [comment]
Removed all node storage classes for use gson: https://github.com/wheremyfoodat/Panda3DS/pull/348/commits/71109d701d37fd7117768fb1c41c0718dc0b5db4

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/lang/Function.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.panda3ds.pandroid.lang;
```
Comment (OFFTKP) [inline src/pandroid/app/src/main/java/com/panda3ds/pandroid/lang/Function.java:1]: What is this class for? Seems unused. Same as gson, if it's for a future PR, add it in the future PR.

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        } else if (value instanceof Float) {
+            data.edit().putFloat(key.name, (Float) value).apply();
+        } else {
+            throw new IllegalArgumentException("Invalid global config value instance");
+        }
+    }
+
+    private static class Key<T extends Serializable> {
+        private final String name;
+        private final T defValue;
```
Comment (OFFTKP) [inline src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java:54]: I don't like abbreviations in variable names. `defValue` made me think definedValue at the beginning, until I googled and realized it's defaultValue. Use the name defaultValue instead so it's immediately obvious what this is.

File: src/pandroid/app/build.gradle.kts [hunk 3]
```diff
@@ -39,4 +39,5 @@ dependencies {
     implementation("androidx.appcompat:appcompat:1.6.1")
     implementation("com.google.android.material:material:1.8.0")
     implementation("androidx.constraintlayout:constraintlayout:2.1.4")
+    implementation("com.google.code.gson:gson:2.10.1")
```
Comment (OFFTKP) [inline src/pandroid/app/build.gradle.kts:42]: gson seems unused for now. PRs should only contain code that is relevant, not code for future changes.

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            value = data.getBoolean(key.name, (Boolean) key.defValue);
+        } else if (key.defValue instanceof Long) {
+            value = data.getLong(key.name, (Long) key.defValue);
+        } else {
+            value = data.getFloat(key.name, ((Number) key.defValue).floatValue());
+        }
+        return (T) value;
+    }
+
+    //Need synchronized why SharedPreferences don't support aysnc write
```
Comment (OFFTKP) [inline src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java:35]: No need for this comment

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+public class GlobalConfig {
+    private static SharedPreferences data;
+
+    public static void initialize() {
+        data = PandroidApplication.getAppContext()
+                .getSharedPreferences(Constants.PREF_GLOBAL_CONFIG, Context.MODE_PRIVATE);
+    }
+
+    private static <T extends Serializable> T get(Key<T> key) {
```
Comment (OFFTKP) [inline src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java:19]: I might be mistaken, but shouldn't these get/set static methods be public? So they can be accessed from other classes? Or do we not need that?

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        Serializable value;
+        if (key.defValue instanceof String) {
+            value = data.getString(key.name, (String) key.defValue);
+        } else if (key.defValue instanceof Integer) {
+            value = data.getInt(key.name, (int) key.defValue);
+        } else if (key.defValue instanceof Boolean) {
+            value = data.getBoolean(key.name, (Boolean) key.defValue);
+        } else if (key.defValue instanceof Long) {
+            value = data.getLong(key.name, (Long) key.defValue);
```
Comment (OFFTKP) [inline src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java:28]: Is there a reason we use the object types (`Long`, `Boolean`) on some cases and the primitive on others (`int`, `float`). If possible let's try to be consistent and use the primitive types for the numbers (`int`, `long`, `boolean`, `float`). `String` is fine. 

For example, looking at the signature of getLong: `long getLong(String key, long defValue)` it returns `long` and not `Long`

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        } else {
+            value = data.getFloat(key.name, ((Number) key.defValue).floatValue());
```
Comment (OFFTKP) [inline src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java:30]: Assuming my other comment is true that we can use primitive types, is there really a need to cast to `Number` first? can't we just `(float)key.defaultValue`?

**Review by OFFTKP — CHANGES_REQUESTED** [review]

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java (hunk 6 above)
Comment (GabrielBRDeveloper) [reply src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java:28]: i know, but java dont support instanceof for primitives: 
image

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java (hunk 7 above)
Comment (GabrielBRDeveloper) [reply src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java:30]: this really not necessary i'll remove this

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java (hunk 5 above)
Comment (GabrielBRDeveloper) [reply src/pandroid/app/src/main/java/com/panda3ds/pandroid/data/config/GlobalConfig.java:19]: True, this really should be public

File: src/pandroid/app/src/main/java/com/panda3ds/pandroid/lang/Function.java (hunk 1 above)
Comment (GabrielBRDeveloper) [reply src/pandroid/app/src/main/java/com/panda3ds/pandroid/lang/Function.java:1]: Ok

**Comment by GabrielBRDeveloper** [comment]
Resolved in: https://github.com/wheremyfoodat/Panda3DS/pull/348/commits/cc44801acdb59887ca88c069e85b85e11912a548

**Review by OFFTKP — APPROVED** [review]

**Comment by OFFTKP** [comment]
lgtm

**Review by wheremyfoodat — APPROVED** [review]

**Pull request merged**